### DISCUSSION TITLE
[Feature Request]:  Generate node ID out of public key

### DISCUSSION THREAD

**Discussion by contra-bit** [meshtastic/firmware#5007] [Ideas]
### Platform

Cross-Platform

### Description

I'd like to propose creating the node ID (aka node number) out of the public key, similar to how software like Tor and SSH handle it. 
This would allow users to choose whether a factory reset node should appear as new or retain its old identity.

Currently, the node ID is generated from the MAC address, which leads to issues when a node is reset and a new public and private key pair is generated without warning. 
This causes PKI mismatches and prevents functions like updating the position or getting a new node name.

By generating the node ID from the public key, users can choose to back up their private and public keys and restore them after a factory reset, allowing the device to retain its identity on the network. 
If the keys are not restored, the device will appear as a new node.

This approach also solves a a privacy concern, as it allows users to appear as a new device on the network if they choose to.

It's essential to inform users about the importance of backing up their keys, as mentioned in Feature Request [#272](https://github.com/meshtastic/web/issues/272). 
This should be done in a secure and user-friendly way, such as prompting the user to download or print a copy of their private keys.

Regarding the open points in Feature Request [#272 Prepare for PKI & Config Updates](https://github.com/meshtastic/web/issues/272):

> Display message when others in conversation change their key.

Tying the node ID the public key, will minimize the risk of a user resetting their node.
The chance of a user accidentally messing up their public key is minimized.

> Allow user to forget node, allowing it to rejoin with new key

This can be implemented by adding an option to "forget" a node, which would remove the node's public key and any associated data such as the node id from the local device. 
Rejoining the network, with a different public key and the same node id, will not be possible.
This will be the same as deleting a node. If the user wants to keep the node, he will restore his keys.

Here's a possible implementation:
- When generating a new node ID, use a hash of the public key instead of the MAC
  address.
- Provide an option for users to choose whether a factory reset node should retain its old identity or appear as new. (This would create a backup of the private and public keys and update the node, after the factory reset).
- Document the process of adding multiple remote management keys and the implications of adding only the key of a remote device.
- Implement notifications for key changes in conversations.
- Implement the "forget node" option.

This feature would improve the overall security, privacy, and user experience of the Meshtastic network. It addresses the issues mentioned in Feature Request [#272 Prepare for PKI & Config Updates](https://github.com/meshtastic/web/issues/272) and [#281 Prompt User to download PKI keys for backups](https://github.com/meshtastic/web/issues/281), and provides a more robust and user-friendly PKI system.
   
## Relevant issues
- [#281 Prompt User to download PKI keys for backups](https://github.com/meshtastic/web/issues/281)
- [#272 Prepare for PKI & Config Updates](https://github.com/meshtastic/web/issues/272)
- [#315  Add ability to add all 3 PKI admin keys](https://github.com/meshtastic/web/issues/315)
- [Key Exchange not mentioned yet](https://github.com/meshtastic/web/issues/281)
- [Meta PKI Issue](https://github.com/meshtastic/firmware/pull/1509)
- [Dicussion on the meaning of Node ID](https://meshtastic.discourse.group/t/clarification-on-user-configuration-and-user-ids-and-node-numbers/10681/4)
- [Relevant for how to implement the Node ID generation](https://github.com/meshtastic/firmware/pull/3651)

**Comment by garthvh** [comment]
Seems likely to cause more issues, we are pretty far down the path of using the BLE Mac, there is a way to reset this on all platforms if you really want to.

**Reply by contra-bit** [reply]
@garthvh Could you please detail what specific issues switching to a node ID generated out of the public key would create? I am wondering if they could be easily addressed.

I must stress that the node ID now got intertwined with the PKI infrastructure introduced by @jp-bennett (after @edinnen work). Keeping the node ID tied to the BLE MAC is already creating issues, such as the pub key mismatch after a factory reset that I mentioned (which makes all nodes that had the previous pub key unable to talk with it).

Forcing the user to change the devices' BLE MAC addres in order to solve this (to get a new node ID) seems quite bad UX.

In order to address issues such as this, switching to a pubkey derived node ID, should keep the code base more simple in the near-future. (That's why projects with PKIs opt for that.)

P.S. Could you please link me to how I may generate a new BLE MAC address for the supported Meshtastic devices?

**Reply by garthvh** [reply]
Those mismatches are intentional, telling you that the node you are communicating with may have been compromised. There are a lot of advantages to using a hardware bound number, using a generated id like the pki is going to substantially increase spoofing and reduce the effectiveness of blacklisting nodes.

**Reply by GUVWAF** [reply]
Why do you want to generate a new BLE MAC address? I don't even think that's possible, it's usually burned in by the vendor (and e.g. for the RP2040 that doesn't have BLE it's a unique identifier). Upon a public key mismatch, you can delete the node from your list (if you want, after verifying with the owner that he did a factory reset) and it will update to the new public key automatically.

> Could you please detail what specific issues switching to a node ID generated out of the public key would create? 

Generating a new NodeNum every time you factory reset causes your node to be duplicated in everyone's node list, and they wouldn't know who to send to. It also increases the chance of duplicate NodeNums.

**Reply by contra-bit** [reply]
Thanks @GUVWAF and @garthvh for all the details and for bearing with me. Please allow me to better elucidate some points.

First of all, let's keep in mind that Meshtastic is a permissionless network, anyone can join or leave at any time. There's no access control.

> Generating a new NodeNum every time you factory reset causes your node to be duplicated in everyone's node list

It wouldn't be a duplicate. There would be no way to tell it's the very same device. It would be as if it was a new node entering the network. This is exactly what's lacking. It is a desirable privacy and anonymity feature.
If the user would like to prevent that (i.e. keep the same ID) it would be just a matter of restoring the keys from a backup.

> they wouldn't know who to send to.

That's the point. There would be a new node in the mesh whom others may try to contact, and there would be a node that went offline (forever).

Indeed this would increase the network churn. Is the size of the nodes list such a big concern? If so, I think it should be easy to alleviate by removing old nodes (not seen in the last X days).

> Upon a public key mismatch, you can delete the node from your list

You should look at the problem the other way around.
As is, whenever someone does a factory reset (and forgets to backup the keys), he ends up with a hardware device that is now useless for the local Meshtastic network. It can no longer send messages or transmit telemetry data (battery level, GPS coordinates, etc.). There's even the chance of creating a network partition.

Being forced to ask (out-of-band) all other node operators (sometimes dozens) to delete your node from their list, in order to fix this, does not align with the permissionless nature of Meshtastic.

Having the option to get into the network with a new ID would fix this.

However, if you wish to retain your previous ID, you could still do it, by just restoring your keys.

> Why do you want to generate a new BLE MAC address?

To get a new ID attributed to a node, so it may appear as a new fresh node on the network.
This way I won't have to bother dozens of node operators for something that was just a test. (The alternatives right now are either to go buy new hardware or modify the ID generation code.)

> Those mismatches are intentional, telling you that the node you are communicating with may have been compromised. 

With a node ID corresponding to the pubkey (or a hash of it) you could still have similar warnings. You could check if the triplet (node ID, short name, long name) matches what is in the node DB, and warn the user in case there's a mismatch. This is akin to what SSH does: it checks whether hostname matches the recorded pubkey.

> using a generated id like the pki is going to substantially increase spoofing

No. It's actually the opposite. If correctly implemented it will eradicate ID spoofing. A new node would be inserted into the local DB only after checking that it provides a valid signature of its node info. With this, an attacker could only spoof if he was able to crack someone's private key, which is supposed to be (nearly) impossible.

Whereas, with the current implementation, an attacker can just edit the source code and spoof any device ID at will. He can even go further and try some social engineering to request known node operators to forget the previous keys (sure this can be mitigated through secure communication and proper verification processes).

> and reduce the effectiveness of blacklisting nodes

Perhaps. But I would argue it is quite ineffective already. A potential spammer can just change the source code and generate new IDs at will.
(The only real solution for this would be a 180 degrees change to the security model of Meshtastic: having a whitelist instead of a blacklist. A node would only communicate with what its operator allows. But this is undesirable on a mesh due to connectivity reasons.)

> we are pretty far down the path of using the BLE Mac

This, i.e. retro-compatibility, seems to me as the major hindrance to my proposal. All existing nodes would change ID upon update.

IMHO the long-term usability, privacy and security advantages far outweigh the temporary disruption that it may cause (which can be alleviated by proper documentation and warnings accompanying the release). However, if the disruption caused is deemed too serious, a migration path may be devised. For instance: allowing nodes to keep the old MAC-based ID for a few releases and letting operators issue the migration themselves when ready.

I believe that the sooner this is done the better. Keeping a PKI infrastructure coexisting with a non-cryptographic ID will only create more usability, privacy and security issues as time goes on.

**Reply by GUVWAF** [reply]
> First of all, let's keep in mind that Meshtastic is a permissionless network, anyone can join or leave at any time. There's no access control.

Only if you use the default channel key, you can also use only private channel keys. 

> It wouldn't be a duplicate. There would be no way to tell it's the very same device. 

There are other human identifiers associated with the node, such as the long and short name (which you usually keep the same). 

> Is the size of the nodes list such a big concern? If so, I think it should be easy to alleviate by removing old nodes (not seen in the last X days).

Yes it is, and removing after X days requires the node to have a notion of time, which is not always the case. Or, it requires it to be always online so it can use relative time over a long period.

> As is, whenever someone does a factory reset (and forgets to backup the keys), he ends up with a hardware device that is now useless for the local Meshtastic network. It can no longer send messages or transmit telemetry data (battery level, GPS coordinates, etc.). 

No, sending messages to a channel and transmitting telemetry just works, because that uses the channel key. 

> Being forced to ask (out-of-band) all other node operators (sometimes dozens) to delete your node from their list, in order to fix this, does not align with the permissionless nature of Meshtastic.
Having the option to get into the network with a new ID would fix this.

With a new ID, you have to find out which node that originally was, so you know who is sending to you and thus which node you need to contact from now on (and then delete the old node to avoid confusion). In my opinion, that’s an even worse UX. 

> Whereas, with the current implementation, an attacker can just edit the source code and spoof any device ID at will.  

Yes, and that’s why you have to verify with the owner whether the public key change was really intended. With your approach, you can also generate new IDs and say “Hey I factory reset my node, contact me here now”.

> A node would only communicate with what its operator allows. But this is undesirable on a mesh due to connectivity reasons. 

You can just use only private channels, packets are still routed via other nodes as long as LoRa settings match (and the nodes didn't explicitly disable this feature).

>  Keeping a PKI infrastructure coexisting with a non-cryptographic ID will only create more usability, privacy and security issues as time goes on.

I’m curious about the privacy and security issues. Also would like to hear what @jp-bennett thinks of this proposal in general.

**Reply by randshell** [reply]
@GUVWAF As a side note on the Bluetooth privacy topic, there seems to be support for random MACs in the protocol itself: https://duo.com/decipher/understanding-bluetooth-security.

I'm note sure if it can be useful for Meshtastic 3.0, but I'll leave it here just in case.

**Reply by GUVWAF** [reply]
Yes, I also know that some phones do this for the WiFi MAC address as well. But I'd argue this is a nice-to-have (since there are other ways to fingerprint radios as well); not a necessity and it comes with downsides.

What makes it difficult for Meshtastic is that: 
- Every time someone forgets to back up its keys it will generate a new entry in everyone's NodeDB and they have to somehow figure out that those are actually the same nodes and one should be deleted, because it cannot be addressed anymore with this ID. 
- Even if you back up your keys, the nodes send out a NodeInfo packet almost immediately after booting, which thus also results in a stale entry in everyone's database every time you factory reset a node.

**Reply by randshell** [reply]
> What makes it difficult for Meshtastic is that:

The random MAC I wrote about is specifically about the Bluetooth protocol and so the communication with, for example, the Android device.

**Reply by GUVWAF** [reply]
> The random MAC here is specifically about the Bluetooth protocol and so the communication with, for example, the Android device.

Okay, I see what you mean. However, OP wanted to change the Bluetooth MAC since that is used to generate the Node ID, not for Bluetooth-specific privacy in general.

**Reply by randshell** [reply]
Oh, all right. I thought it was a request for an additional privacy feature, but I guess it was just a case of the "XY problem". :slightly_smiling_face: 

Btw. Thanks again for all the support.

**Comment by contra-bit** [comment]
Thanks (again :) ) @GUVWAF for giving me more details.

There's one important bit I think you're missing: one of the major advantages of my proposal is to allow a node to get a totally new identity, unrelated to the previous one.

Thus:

> There are other human identifiers associated with the node, such as the long and short name (which you usually keep the same).

You wouldn't keep them the same. That's the point.

> With a new ID, you have to find out which node that originally was, so you know who is sending to you and thus which node you need to contact from now on

This would not happen too. For other node operators it would be just as if a new node appeared on the mesh. They don't know whom it belongs to. And that's what is intended.

There a plenty of reasons for a user wishing to get a new ID for his node. Examples:
- You're selling/giving it to someone else.
- You want to repurpose the node for another role (e.g. turn it from your mobile client to a remote sensor).
- You want to avoid tracking (e.g. you reset the ID when you travel to a different place).

With the introduction of PKI, this need became even more pressing, as now the node ID gets associated with a specific pubkey on everyone's node DB.

With the current MAC-based forever fixed node ID, stuff like repurposing a node or giving it to a friend, which before the PKI could be just a matter of doing a factory reset and setting a new node name, now involves coordination with all the peers of your mesh. That's lots of effort for no gain, as you do not wish (in all these scenarios) the ever again "be the same".

On the other hand, if you do wish the keep the previous ID, it will be just a matter of backing up your keys. Easy.

Now, on the finer details:

> [Permissionless] Only if you use the default channel key

Not only that. A new node is immediately recorded to the node DB and used for relaying messages. No barriers to join the mesh.

> Yes it is, and removing after X days requires the node to have a notion of time, which is not always the case. 

Ohhh… So there's no automatic way to prune the node DB? What happens when it becomes full? The node just stops accepting new peers?

If so, that's quite an attack vector. It should be addressed anyway.

> No, sending messages to a channel and transmitting telemetry just works, because that uses the channel key.

Indeed, it works for channels, as those use a PSK. But it stops working for direct communication.

I tested it. I was giving a node (that I just used for tests) to a friend. I did a factory reset (without saving the PKI keys of course), and now all peers are no longer able to read the node's battery level and GPS coordinates (due to the ID-to-key mismatch).

We are now forced to contact everyone in the mesh whom already runs the PKI-enabled version. It could have been much simpler if the factory reset also gave it a need ID: my friend would now have a fresh device that had no correlation with my previous usage.

> With your approach, you can also generate new IDs and say “Hey I factory reset my node, contact me here now”.

Sure, but that attack will become harder to pull off.

With a cryptographically bound ID, the attacker won't be able to spoof the ID, only the node name. The victim receiving that message will thus see 2 nodes online with the same name, but with different IDs.

Right now, things are worse: the attacker can spoof the ID too! The victim will only see 1 node on his list.

----

> I’m curious about the privacy and security issues

To sum it up simply, the advantages of using an ID bound to the pubkey, instead of a fixed one based on the MAC address, are:

- Privacy/anonymity: the same hardware can now get a different ID, and thus be repurposed/sold/offered/wtv without tying it to the previous usage.
- Security: spoofing IDs becomes impossible (or rather, super computationally hard).

**Reply by GUVWAF** [reply]
I’m sorry, but what has getting a new identity for re-purposing a node to do with any of this? To re-purpose a node, just give it a new long and short name. The old node doesn’t exist anymore, so it doesn’t make sense to leave multiple nodes in someone’s NodeDB. There are position precision settings to avoid tracking, or just disable it on the public channel.

The point was that you get a public key mismatch if you don’t back-up your keys when doing a factory reset, and nodes on the receiving end need to delete that node before being able to send DMs again. You proposed to generate a new Node ID in that case and then: 

> They don't know whom it belongs to. And that's what is intended. 

Why is this intended after only a factory reset? **Edit**: I may be able to see a use-case for generating a completely new ID (although I would not be in favor of it because it just leads to stale entries in people's NodeDBs), but that shouldn't be the default for when you factory reset. 

> Not only that. A new node is immediately recorded to the node DB and used for relaying messages. No barriers to join the mesh.

You can’t send or receive messages to/from those nodes if you don’t share a channel key - that’s not really joining the mesh. Also, nodes without NodeInfo will be removed from the DB when rebooting. 

> What happens when it becomes full? The node just stops accepting new peers?

It does try its best to remove the oldest not favorite node. 

> I tested it. I was giving a node (that I just used for tests) to a friend. I did a factory reset (without saving the PKI keys of course), and now all peers are no longer able to read the node's battery level and GPS coordinates (due to the ID-to-key mismatch).

Then something else must have gone wrong. Battery level and GPS coordinates are not using PKI, because they are broadcast to a channel.

> Right now, things are worse: the attacker can spoof the ID too! The victim will only see 1 node on his list. 

Yeah, but it can’t decrypt packets encrypted with the public key of the old node, and you only update the public key after you delete the node. With your proposal, if you accidentally choose the node with wrong NodeID, you may send private information to it.

**Comment by contra-bit** [comment]
> what has getting a new identity for re-purposing a node to do with any of this?

Because with the new PKI infrastructure the node ID now got tied to the node's pubkey.

> To re-purpose a node, just give it a new long and short name. The old node doesn’t exist anymore

That no longer suffices. The old node still exists in everyones' node DB and is associated with the previous pubkey. Since, with the current code base, the new node re-uses the previous node ID you will get a key mismatch.

Thus, if for instance, a new person wants to use this new node, he will have to reach all peers that he has in common with the previous owner and say "hey, please delete the node, it's now owned by a different person".

> Why is this intended after only a factory reset? 

That's what factory resets do everywhere else, they delete the previous owner state. IMHO the node ID should be part of that state. Take a look at how it works for a smartphone.

Anyway, the default can be to keep the node ID and cryptograhic keys.

> There are position precision settings to avoid tracking

Tracking comes in many different forms. Not only listening to the node's position updates.
Even if a node sends no position information, it's still possible to know its rough position, just by knowing whom it is connecting to, the radio signal power, etc. That's how mobile phones are tracked accross the cell phone networks.

Since, as of now, the node ID can never change, it's easy to track a device accross the network. An attacker can know that you went from one region to another, or from one country to another. (With messages being relayed over MQTT the attacker may not even need to have physical nodes in all places).

For some reason or another you may wish to not reveal you're using Meshtastic on another place. The easy solution would be to get a new node ID, but that's impossible now (without change the source code).

> Battery level and GPS coordinates are not using PKI, because they are broadcast to a channel.

Indeed. I just re-tested it again and I was able to get the battery level and GPS coordinates. The only thing that did not work were direct messages (as expected). Sorry for the wrong info on this one, and thanks for poiting it out.

> You can’t send or receive messages to/from those nodes if you don’t share a channel key - that’s not really joining the mesh.

Even without sharing a channel key other nodes will nevertheless retransmit your messages, right? I.e. their resources (in this case power and connectivity) are offered to you with no barrier. That's permissionless in my view.

**Reply by garthvh** [reply]
You can change the node id on every platform if you want to, this is a tiny edge case and there is a solution already that complexity does not need to be applied to every device and user.

**Reply by contra-bit** [reply]
> You can change the node id on every platform if you want to,

Ohh, that's great. Could you please link me somewhere describing how to do it?

**Reply by GUVWAF** [reply]
Indeed, while there may be very niche use-cases where you'd like to change your NodeNum, when adding it as an option to everyone it would only lead to misuse (either intentional or not). Especially making it the default when doing a factory reset will lead to issues.

> Could you please link me somewhere describing how to do it?

You can set one here:
https://github.com/meshtastic/firmware/blob/7fd1c334d3c1390ce8d7beff25cc81e0c0e0ec6e/src/mesh/NodeDB.cpp#L679
Then factory reset and it should pick this one.

>  a new person wants to use this new node, he will have to reach all peers that he has in common with the previous owner and say "hey, please delete the node, it's now owned by a different person"

Re-purposing a node is really a niche use case, and in my view this is a good requirement, because it's a sign the old node in the list does not exist anymore, and is replaced by this specific one. If you change your phone number, you also have to tell everyone that.

> Take a look at how it works for a smartphone.

You mention "_That's how mobile phones are tracked accross the cell phone networks_", but surely not every smartphone gets a new Wi-Fi/BLE MAC address and IMEI after a factory reset. And indeed even without unique identifier you can fingerprint devices by other means.

> I.e. their resources (in this case power and connectivity) are offered to you with no barrier.

Right, they offer *you* resources, but if you set your rebroadcast mode to `KNOWN_ONLY`, you don't offer them any. You can capture and replay any wireless signal, but that doesn't give you "access" to the network.

**Reply by randshell** [reply]
I second this request. But other than privacy reasons, if I lose the keys then am I stuck between compiling the firmware myself and asking for a refund?

**Reply by GUVWAF** [reply]
If you want a new identity, yes. Else you have to tell your contacts you got new keys, so they have to forget your node and let them rediscover.

**Reply by randshell** [reply]
Thank you for the quick reply.
Is there any logic at the moment that lets the device forget me automatically, for example over time? I don't think it's possible to tell everyone in the city, which you don't even know, to forget your device so that you can reach them again.

Let me explain my current situation better. I've got my first Meshtastic device, a T-Echo, I've flashed the reset file, flashed the firmware and met the various nodes of my city. All good.

After playing a bit with it, I've realized that flashing the reset file doesn't reset the private key (I'm not sure about the Bluetooth pairing yet), which is an issue. This means that potentially the manufacturer or the seller could read my conversation, assuming they reach me or a hop connected to me.

I am a privacy enthusiast and it would be great to have IDs tied to the Public Key, like @contra-bit suggested, but at the moment my problem is a slightly different one. If I were to reset the device with the CLI tool and with `--factory-reset-device`, I'd get new keys but I wouldn't be able to talk to "strangers" in my city. In other words, what can I do?

On a bigger topic, is it possible for the reset file to also reset the keys? But then, if someone sells their devices (I'm sure no one gives up on Meshtastic as it's quite addictive but it can definitely happen), the buyer won't be able to communicate to anyone that previously saw this node.

**Reply by GUVWAF** [reply]
> Is there any logic at the moment that lets the device forget me automatically, for example over time?

At the moment, no, unless it cycles out of their NodeDB automatically when it reaches the maximum size.

> I don't think it's possible to tell everyone in the city, which you don't even know, to forget your device so that you can reach them again.

You also don't necessarily need to tell them; they will get a red lock symbol in the app that indicates a key mismatch, after which they can decide themselves whether they trust this is not a malicious actor trying to impersonate the node.

> I'd get new keys but I wouldn't be able to talk to "strangers" in my city. In other words, what can I do?

Also note that PKI is only used for direct messages, you can send a message to a channel saying you factory reset your node.

>  I've realized that flashing the reset file doesn't reset the private key

Hmm, I was under the impression that it did. At least the "Full Erase and Install" option in the flasher does.

**Reply by esev** [reply]
I've modified the firmware on my EDC device as mentioned in a previous comment to set a custom node id. I do this so I can upgrade devices without changing IDs. And so that I can gift a non-used device without transferring the node id.

If there was an **option** to generate the node id from the public key, that would be useful to me as I wouldn't need to maintain a separate firmware branch for this purpose.

Edit: There may be complications with HAM/licenced mode here, as they might not have public keys.

**Reply by thebentern** [reply]
For 3.0 where we have an opportunity to re-work the protocol, I'd almost argue the public key could be considered the new identifier for a node in some sense.

**Reply by randshell** [reply]
> Also note that PKI is only used for direct messages, you can send a message to a channel saying you factory reset your node.

Thank you very much @GUVWAF, that's actually a nice idea without going into rebuilding. :slightly_smiling_face:

> Hmm, I was under the impression that it did. At least the "Full Erase and Install" option in the flasher does.

I'd be open to test it if this turns out to be an issue, just ping me.

---

Hi there @esev, @thebentern!

> I've modified the firmware on my EDC device as mentioned in a previous comment to set a custom node id. I do this so I can upgrade devices without changing IDs. And so that I can gift a non-used device without transferring the node id.

I think configuring Github CI, starting from the already existing yaml configs, should be pretty straightforward actually. Perhaps I could try this approach as well. :thinking: 

> For 3.0 where we have an opportunity to re-work the protocol, I'd almost argue the public key could be considered the new identifier for a node in some sense.

That's great to know, thanks!

**Reply by esev** [reply]
It should persist across upgrades, yes. And my experience is that it needs a factory reset (edit: maybe just a node db reset, I can't remember) to pickup the new ID. If there is ever an issue that requires a factory reset though it'll revert back to the original Mac address node id unless the custom build is in place.

I found the line that @GUVWAF mentioned to be the simplest place to do the firmware modification. Change `nodeNum` to the value you want to use: https://github.com/meshtastic/firmware/blob/191ca8ce124d9adbc9843ec5de0a3edf3f2c0af9/src/mesh/NodeDB.cpp#L902

You need to be careful to not generate a node ID that is common as well. It is tempting to use [hex words](https://nedbatchelder.com/text/hexwords.html), but that'll eventually result in conflicts.

**Reply by contra-bit** [reply]
> a new entry in everyone's NodeDB
> stale entry in everyone's database

It seems the limited node database is the issue. Trying to solve it by limiting IDs at the cost of privacy and usability seems the wrong way to go.
I guess more effort should be put into the NodeDB.

**Reply by GUVWAF** [reply]
Not only the limited size of the NodeDB, but also that the user does not have a way to know that a node is not reachable anymore, or even whether it is a new physical node or just someone factory resetting its device. If you don't have a reliable time source you don't even know which is the most recent heard node with a certain name.

I would be fine with a configuration to generate a new random node number on request, but in my opinion it should not be the default when factory resetting your node.

**Reply by quentinmit** [reply]
What is the use case for keeping a node ID while generating new PKI keys? Any time the keys are reset, the node ID should be reset as well. I like the OP's suggestion of generating the node ID from the public key because it means you can easily migrate a node ID to a new device just by moving the private key.

**Reply by Neo-Desktop** [reply]
what if there was a way to announce that the current node is intending to change its ID? online nodes would then have the option to automatically delete it (or flag it as changed for manual review) and then the node simply changes ID and announces itself with its new identity.

It's like a final sign-off type message essentially

**Reply by jedahan** [reply]
in other public-key cryptogaphy based peer-to-peer systems (like scuttlebutt for instance, and i think theres something similar for activitypub based things) theres a mutually signed message. 
You post 'My new public key will be XYZ', signed by your current public key (ABC).
You then post 'My previous public key was ABC', signed by your new public key (XYZ).

Of course this is more difficult with only a single device, and if you have the ability to backup your current key, then you can migrate to new device anyway without changing keys. But it also allows for multi-node single identity and other interesting properties.

A simpler one would be a key revocation message, but the problem with announcing future intent with no reciprocal is that its even easier for folks to race to takeover previous identity.

**Comment by shortwavesurfer2009** [comment]
This also touches on spam because currently, from what I understand, you can set a node to be ignored by its ID. So if some annoying spammer tries to spam the mesh, routers could just put them in their ignore list. If you have the ability to easily change IDs, this would become a lot less effective. In amateur radio, for example, something like DMR or Yasu System Fusion is much less likely to have people keying up and playing fart sounds, etc. over repeaters, because with DMR you need an ID, etc.

**Reply by sendingstrings** [reply]
Mesh spam isn’t really an issue if you get key locked out of traffic with other nodes….

**Comment by ph1l** [comment]
Let's say I *just* dropped ~$100 on hardware for this project. I plug it in *right away* because I'm excited to participate in the bay area mesh. I'm new, so I make mistakes. I find that everyone moved over to MedumSlow, and join them. for a couple of days this is working great. Then I hear my key *might* be insecure so I regenerate it to be secure.

Suddenly no one can send or receive encrypted messages with me. My packets don't show up on the local website I've been using to debug my connection and the network. Now the entire mesh I want to communicate with has an incorrect key for my node.

Does that original key time out in other nodes? How long will my device be ghosted by nodes that have my original key?

**Reply by sendingstrings** [reply]
Exactly.

**Reply by Terramoto** [reply]
nodedb nodes rotate, favorites are kept.
Could allow for custom node removal timer or share nodeinfo broadcast timer to remove non-favorites.

**Comment by LordSexy** [comment]
I startend and was a big dumbdumb.
Edited the settings until nothing worked, reseted my node to factory settings.
Got a message about my pubkey being already used.
Paniced
Regenerated keys
Can only communicate in the Fast long channels

**Comment by jp-bennett** [comment]
This is in discussion for possible implementation in 2.8.

**Comment by tylerpieper** [comment]
I would also like to voice my opinion on this. I think Node ID from Public Key is not a great move. 

- When you buy a used laptop, you don't get a new MAC address
- Ignoring nodes would be impossible (since there's nothing tied to the hardware)
- Flash corruption (such as from a battery brownout or ESD or cosmic ray) could lead to loss of identity completely. With current Node ID from MAC, at least the Node ID is the same no matter what. 
- PKC is only for DM/admin messages. It does not apply to channel traffic or general LoRa behavior. With default settings, a node will relay a message from a node it has never heard from with a payload it cannot decode. It doesn't need nodeinfo or the correct public key for a node. You can send messages in the public channel and other nodes will receive it even if they have you flagged for a public key mismatch. If they want to DM you, they can choose to delete your nodeinfo to get it to refresh the key. 
- You could even choose to not have the public channel ever configured and just use private channels for everything. Other nodes with the same modem preset and frequency will still relay your traffic just the same (assuming default rebroadcast settings).

For me, the negatives outweigh any positives (and the positives are quite small).

**Reply by quentinmit** [reply]
> I would also like to voice my opinion on this. I think Node ID from Public Key is not a great move.
> 
> * When you buy a used laptop, you don't get a new MAC address

Actually, you do. [iOS, macOS](https://support.apple.com/guide/security/privacy-features-connecting-wireless-networks-secb9cb3140c/web), [Android](https://source.android.com/docs/core/connect/wifi-mac-randomization-behavior), [Windows](https://learn.microsoft.com/en-us/answers/questions/5609414/how-do-i-disable-random-mac-addresses-for-ethernet), and [Linux](https://wiki.archlinux.org/title/NetworkManager#Configuring_MAC_address_randomization) all use random MAC addresses for each connection now. The hardware MAC address is ~unused for anything meaningful.

> * Ignoring nodes would be impossible (since there's nothing tied to the hardware)

It's already impossible. A spammer can already build a custom firmware with whatever node ID they want. All the current behavior does is reduce privacy for average users.

> * Flash corruption (such as from a battery brownout or ESD or cosmic ray) could lead to loss of identity completely. With current Node ID from MAC, at least the Node ID is the same no matter what.

There's already a facility to back up private keys - if you care about a stable identity, just use that.

> For me, the negatives outweigh any positives (and the positives are quite small).

Strong disagree. The existing behavior is both a loaded footgun (because it breaks when a user factory resets their node without backing up keys) and a privacy risk (because a user who factory resets won't know they they can still be tracked across the reset).
